### ISSUE TITLE
KAS ECC/FFC VAL test: key verification requirement

### ISSUE THREAD

**Issue by smuellerDD** [usnistgov/ACVP#768]
With ACVP, the VAL test type of KAS FFC/ECC introduced the requirement that the provided public/private key pair must be validated by the IUT prior to calculating the shared secret compliant to SP800-56A full/partial valiation definitions in section 5.6.

I am not disputing that this is a good thing.

However, suppose you have an IUT that can do full (EC)DH operation for, say, TLS. That IUT only supports one mode of operation: it will always generate the DH keypair for the TLS operation. In order to perform a DH operation, only the public keys are exchanged. As the IUT does not accept a private key from outside, the IUT does not have a key pair check following the full/partial validation of 56A. From a crypto point of view, this is fully appropriate, because the DSA/ECDSA key generation mechanism ensures that the key will always be compliant. I.e. in this scenario, the IUT does not need such a key check mechanism.

To facilitate the ACVP testing, the module provides an interface to import a private key. This interface is solely for testing purpose and is not intended for production use.

Yet, the module now fails the ACVP testing due to the missing key check that the imported private/public key pair is valid.

Bottom line: I think you see that the IUT does not have the key check and does not need it. Yet, the ACVP server insists on that key check.

How can we solve that conundrum?

**Comment by smuellerDD** [comment]
Looking at the ACVP specification, I see the following function types defined (among others)

- FFC: keyPairGen vs fullVal/partialVal - see section 3.3

- ECC: keyPairGen vs fullVal/partialVal - see section 3.3

When I read that correctly, I would infer that the full/partial validation is only verified by the ACVP server if partial/full Val is selected. When keyPairGen is selected, I would assume that the key validation is NOT tested.

Note, my IUT fails on the issue of the key validation as noted before.

Test vectors are requested using the keyPairGen function type. Yet, the generated responses all fail due to the fact that a given private key should have been rejected by the key validation logic.

Bottom line: may I ask that the ACVP server is changed such that the key pair validation is only verified if fullVal/partialVal selected during registration? I.e. when keyPairGen is selected, the key validation is NOT checked?

Thanks

For my IUT, I selected keyPairGen

**Comment by hallnist** [comment]
NIST SP 800-56A rev 3, Section 5.6.2.1.2, at the bottom of p. 35, says:

"The owner of a static or ephemeral key-establishment key pair shall obtain an initial
assurance that the private key is an integer in the correct interval by using one of the
following methods:
a. For both a static and ephemeral key pair: The owner generates the key pair as
specified in Section 5.6.1, or
b. ..."

Is this the scenario you are describing, i.e., the IUT is the owner of the key pair and is claiming (a.) as the initial assurance?

**Comment by smuellerDD** [comment]
Yes, correct. The IUT in production mode is always the owner of the key due to having it created. The IUT never supports an import of the private key. It therefore complies with option (a).

The associated key generation ACVP test is performed successfully.

**Comment by hallnist** [comment]
OK, we are taking the key check out for the case when keyPairGen is selected but full/partial validation is not.

**Comment by smuellerDD** [comment]
Thank you very much.

I would keep the issue open until the change is in if you concur.

**Comment by hallnist** [comment]
Yes, that is fine.

**Comment by Kritner** [comment]
this fix will be in the next release.

**Comment by Kritner** [comment]
This update is now on demo and prod. Let us know if there are any issues
